app/assets/javascripts/controllers/middleware_topology/middleware_topology_controller.js

```diff
@@ -54,66 +56,100 @@ function MiddlewareTopologyCtrl($scope, $http, $interval, $location, topologySer
         $interval.cancel(promise);
     });
 
-    $scope.$on("render", function(ev, vertices, added) {
+    $scope.$on('render', function(ev, vertices, added) {
         /*
          * We are passed two selections of <g> elements:
          * vertices: All the elements
          * added: Just the ones that were added
          */
-        added.attr("class", function(d) { return d.item.kind; });
-        added.append("circle")
-            .attr("r", function(d) { return self.getDimensions(d).r})
+        added.attr('class', function(d) { return d.item.kind; });
+        added.append('circle')
+            .attr('r', function(d) { return self.getCircleDimensions(d).r})
             .attr('class' , function(d) {
               return topologyService.getItemStatusClass(d);
             });
-        added.append("title");
-        added.on("dblclick", function(d) {
+        added.append('title');
+        added.on('dblclick', function(d) {
             return self.dblclick(d);});
-        added.append("image")
-            .attr("xlink:href",function(d) {
-                return "/assets/svg/" + self.class_name(d) + ".svg";
+        added.append('image')
+            .attr('xlink:href',function(d) {
+                var iconInfo = self.getIcon(d);
+                switch(iconInfo.type) {
+                    case 'image':
+                        return iconInfo.icon;
+                    case 'glyph': // the icon will be provided as a text
+                        return null;
+                }
             })
-            .attr("y", function(d) { return self.getDimensions(d).y})
-            .attr("x", function(d) { return self.getDimensions(d).x})
-            .attr("height", function(d) { return self.getDimensions(d).height})
-            .attr("width", function(d) { return self.getDimensions(d).width});
-        added.append("text")
-            .attr("x", 26)
-            .attr("y", 24)
-            .text(function(d) { return d.item.name }).style("font-size", function(d) {return "12px"}).style("fill", function(d) {return "black"})
-            .style("display", function(d) {if ($scope.checkboxModel.value) {return "block"} else {return "none"}});
-
-        added.selectAll("title").text(function(d) {
-            return topologyService.tooltip(d).join("\n");
+            .attr('y', function(d) { return self.getCircleDimensions(d).y})
```

Comment:
Right, this can be fixed by the D3 `each()` function similarly as in here https://git.io/voPqb The code will be more complex though, but the resulting HTML will be better => potential speedup. So it's worth doing